benchmarks/encryption.py

```diff
@@ -0,0 +1,330 @@
+
+    def __init__(self, password, iter, hash, derive, backend, salt):
+
+        self._password = password
+        self._iterations = iter
+        self._hash = hash
+        self._derive = derive
+        self._backend = backend
+        self._salt = salt
+
+        self._length = 16
+
+    def _generate_key(self):
+
+        kdf = self._derive(
+            algorithm=self._hash(),
+            length=self._length,
+            salt=self._salt,
+            iterations=self._iterations,
+            backend=self._backend()
+        )
+
+        return kdf.derive(self._password)
+
+    def _verify(self, k):
+
+        kdf = self._derive(
+            algorithm=self._hash(),
+            length=self._length,
+            salt=self._salt,
+            iterations=self._iterations,
+            backend=self._backend()
+        )
+
+        kdf.verify(self._password, k)
+
+    def __call__(self):
+
+        key = self._generate_key()
+        self._verify(key)
+        return key
+
+
+class Encrypt(object):
+
+    def __init__(self, key, hmac_key, hmac, hash, cipher, algorithm, mode, backend, salt, hmac_salt):
+
+        self._key = key
+        self._hmac_key = hmac_key
+        self._hmac = hmac
+        self._hash = hash
+        self._cipher = cipher
+        self._algorithm = algorithm
+        self._mode = mode
+        self._backend = backend
+        self._salt = salt
+        self._hmac_salt = hmac_salt
+
+        self._iv = urandom(16)
+
+    def __call__(self, data):
+
+        h = self._hmac(
+            self._hmac_key,
+            self._hash(),
+            backend=self._backend()
+        )
+
+        ciph = self._cipher(
+            self._algorithm(self._key),
+            self._mode(self._iv),
+            backend=self._backend()
+        )
+
+        encryptor = ciph.encryptor()
+        iv_sent = False
+
+        for raw in data:
+            out = encryptor.update(raw)
+
+            if not iv_sent:
+                out = self._iv + self._salt + self._hmac_salt + out
+                iv_sent = True
+
+            h.update(out)
+
+            yield out
+
+        out = encryptor.finalize()
+        h.update(out)
+        yield out
+
+        signature = h.finalize()
+        yield signature
+
+
+class Decrypt(object):
+
+    def __init__(self, password, hash, derive, hmac, cipher, algorithm, mode, backend):
+
+        self._password = password
+        self._hash = hash
+        self._derive = derive
+        self._hmac = hmac
+        self._cipher = cipher
+        self._algorithm = algorithm
+        self._mode = mode
+        self._backend = backend
+
+    def _derive_key(self, salt):
+
+        key_derive = KeyDerive(
+            self._password,
+            PASSWORD_DERIVE_ITER,
+            self._hash,
+            self._derive,
+            self._backend,
+            salt
+        )
+
+        return key_derive()
+
+    def _return_chunks(self, data, minsize=48):
+
+        cur, ahead = tee(data)
```

Comment:
Is there any way to avoid using tee? From my understand, tee may require quite a bit of overhead. From the [stdlib docs on tee](https://docs.python.org/3/library/itertools.html#itertools.tee):

> This itertool may require significant auxiliary storage (depending on how much temporary data needs to be stored). In general, if one iterator uses most or all of the data before another iterator starts, it is faster to use list() instead of tee().

Reply:
Ah I see. Interesting. I'll see what I can do!